jetty-ee10/jetty-ee10-servlet/src/main/java/org/eclipse/jetty/ee10/servlet/ServletChannel.java

```diff
@@ -525,16 +519,11 @@ public boolean handle()
                             }
                             else
                             {
-                                // TODO: do this non-blocking.
-                                // Callback completeCallback = Callback.from(() -> _state.completed(null), _state::completed);
-                                // _state.completing();
-                                try (Blocker.Callback blocker = Blocker.callback())
-                                {
-                                    // We do not notify ServletRequestListener on this dispatch because it might not
-                                    // be dispatched to an error page, so we delegate this responsibility to the ErrorHandler.
-                                    dispatch(() -> errorHandler.handle(_servletContextRequest, getServletContextResponse(), blocker));
-                                    blocker.block();
-                                }
+                                // We do not notify ServletRequestListener on this dispatch because it might not
+                                // be dispatched to an error page, so we delegate this responsibility to the ErrorHandler.
+                                // We must ignore the callback because the error servlet could go async.
+                                dispatch(() ->
+                                    errorHandler.handle(_servletContextRequest, getServletContextResponse(), Callback.NOOP));
```

Comment:
@lachlan-roberts I don't think this going to work... or at least it should not work and if it does it is probably a race or only working because the error page is small and does not congest the connection.

The continuation of the handling loop needs to be delayed until the callback is called. Something like what was suggested in the TODO comment previously: `Callback completeCallback = Callback.from(() -> _state.completed(null), _state::completed);`    although I don't think that is exactly correct.     It might just need a call to get the handling going again... which might need to be executed?  We might also need an optimisation to see if the callback has been completed in the call to errorHandler.handle, in which case we can continue, if not then the callback has to continue handling.

Reply:
The problem is that in `ErrorHandler` we do:

```java
errorDispatcher.error(httpServletRequest, httpServletResponse);
callback.succeeded();
```

so after callback is succeeded but we could still be asynchronously writing the response with an async dispatch. So in that case of a servlet error page the callback means effectively nothing. So calling `state.Completing()` or `state.completed()` is wrong.

I think this change I made will work for any servlet error page but not a Jetty core one. So maybe we need to do a `scheduleDispatch()` in the callback?

Reply:
Take a look at the further changes I made to this.

Reply:
Yeah, even if we "know" the thing we are calling is currently synchronous, if we are passing a callback, we should strive to call it in correct style.   Your changes are better, but I'll comment inline how to improve.